src/main/java/io/strimzi/kafka/quotas/throttle/PolicyBasedThrottle.java

```diff
@@ -0,0 +1,117 @@
+     * @param listener     the lister to be notified of changes
+     * @param expiryPolicy expiry policy to control how long a factor is applied for
+     * @param fallbackThrottleFactor throttle factor to apply if a factor reaches its expiry
+     */
+    public PolicyBasedThrottle(ThrottleFactorPolicy factorPolicy, Runnable listener, ExpiryPolicy expiryPolicy, double fallbackThrottleFactor) {
+        this(factorPolicy, listener, Clock.systemUTC(), expiryPolicy, fallbackThrottleFactor);
+    }
+
+    /* test */ PolicyBasedThrottle(ThrottleFactorPolicy factorPolicy, Runnable listener, Clock clock, ExpiryPolicy expiryPolicy, double fallbackThrottleFactor) {
+        this.factorPolicy = factorPolicy;
+        this.listener = listener;
+        this.clock = clock;
+        this.expiryPolicy = expiryPolicy;
+        throttleFactor = ThrottleFactor.validFactor(1.0d, clock.instant(), this.expiryPolicy);
+        this.fallbackThrottleFactor = fallbackThrottleFactor;
+    }
+
+    @Override
+    public ThrottleFactor currentThrottleFactor() {
+        return throttleFactor;
+    }
+
+    @Override
+    public void observeVolumeUsage(VolumeUsageResult observedVolumes) {
+        updateFactor(current -> getNewFactor(observedVolumes, current));
+    }
+
+    /**
+     * Check for expired factor and fallback if it has expired.
+     */
+    public void checkThrottleFactorValidity() {
+        log.info("Checking for expired factor");
+        try {
+            updateFactor(this::maybeFallback);
+        } catch (Exception e) {
+            log.warn("Failed to check for stale factor", e);
+        }
+    }
+
+    private void updateFactor(Function<ThrottleFactor, ThrottleFactor> throttleFactorUpdater) {
+        boolean changed = updateFactorAndCheckIfChanged(throttleFactorUpdater);
+        if (changed) {
+            listener.run();
+        }
+    }
+
+    private synchronized boolean updateFactorAndCheckIfChanged(Function<ThrottleFactor, ThrottleFactor> throttleFactorUpdater) {
+        ThrottleFactor currentFactor = this.throttleFactor;
+        throttleFactor = throttleFactorUpdater.apply(currentFactor);
+        boolean changed = !Objects.equals(currentFactor.getThrottleFactor(), throttleFactor.getThrottleFactor());
+        if (changed) {
+            log.info("Throttle Factor changed from {} to {}, notifying listener", currentFactor, throttleFactor);
+        } else {
+            log.debug("Throttle Factor unchanged at {}, not notifying listener", throttleFactor);
+        }
+        return changed;
+    }
+
+    private ThrottleFactor getNewFactor(VolumeUsageResult observedVolumes, ThrottleFactor current) {
+        if (observedVolumes.getStatus() == SUCCESS) {
+            return calculateThrottleFactorWithExpiry(observedVolumes);
+        } else {
+            return maybeFallback(current);
+        }
+    }
+
+    private ThrottleFactor calculateThrottleFactorWithExpiry(VolumeUsageResult observedVolumes) {
```

Comment:
Inlined in 12ebcc11b80028d5f9a4c14007234ece4da934b4

I think we had split it out as it gave some explanation as to what that code was doing and gave it a sense of symmetry with the `maybeFallback` case, but given feedback that it would be clearer inline we have done so.